Add support for DS2760 1-Wire TypeK Thermocouple

droberg opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for DS2760 1-Wire TypeK Thermocouple

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/20986

The DS2760 Li+ Battery Monitor [1] can easily be used as a thermocouple
converter because it has both temperature and voltage reading. 
The use is supported through OWFS  [2]. Unfortunately the original 
app note from Maxim is no longer available online.

[1] https://pdfserv.maximintegrated.com/en/ds/DS2760.pdf
[2] https://owfs.org/uploads/ds2760.html

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @droberg, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#63222).

Once you do that we will be able to review and accept this pull request.

Thanks!

davet2001 reviewed (approved):
Looks good to me but I am unable to test this.

epenet commented:
Can you please update the tests constants file?

homeassistant commented:
Hello @droberg,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`12ecc6b79b0e0d10fbd28d993fdd0b541a1bb990`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;63222&#x2F;commits&#x2F;12ecc6b79b0e0d10fbd28d993fdd0b541a1bb990) - This commit has something that looks like an email address (pero@wolfram.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

droberg committed:
Add test mockup sensor for DS2760 to tests

droberg commented:
> Can you please update the tests constants file?

I have now updated the files requested, but I accidentally messed up the authorship. I thought I fixed this with a forced update but I am unsure whether my fix worked.

garbled1 commented:
I'm good with the code, but I don't claim to understand the wierd cla issues.

epenet commented:
I wonder if we should have a device/entity config setting for the thermocouple type.
This PR is hardwired to use type K, but according to [OWFS documentation](http://owfs.sourceforge.net/DS2760.3.html#sect13) `supported thermocouple types include types B, E, J, K, N, R, S and T`

It would be great to have a configuration drop-down for the type, instead of being hard-wired to K.

droberg commented:
Yes, you are probably right. 
I'll give it a try.

garbled1 commented:
does the name change dependent on type?  or does it always show up as "K" even if it's a "T" ?

Does it even matter?  I mean, do the numbers not work right if you plug in a T?

epenet commented:
> does the name change dependent on type? or does it always show up as "K" even if it's a "T" ?
> 
> Does it even matter? I mean, do the numbers not work right if you plug in a T?

It looks to me like we could add 8 temperature sensors:
- "typeB/temperature"
- "typeE/temperature"
- "typeJ/temperature"
- "typeK/temperature"
- "typeN/temperature"
- "typeR/temperature"
- "typeS/temperature"
- "typeT/temperature"

But I think if we find a way to have just a single thermocouple temperature sensor which is configurable it would be better.

droberg commented:
In OWServer it shows up as all variants at the same time. The difference is that when you read the ".../typeK/temperature" path/catalog or what they call it a different scaling is applied in the voltage to temperature conversion. So all values are wrong except the one matching your actual thermocouple.

droberg commented:
> > does the name change dependent on type? or does it always show up as "K" even if it's a "T" ?
> > Does it even matter? I mean, do the numbers not work right if you plug in a T?
> 
> It looks to me like we could add 8 temperature sensors:
> 
> * "typeB/temperature"
> * "typeE/temperature"
> * "typeJ/temperature"
> * "typeK/temperature"
> * "typeN/temperature"
> * "typeR/temperature"
> * "typeS/temperature"
> * "typeT/temperature"
> 
> But I think if we find a way to have just a single thermocouple temperature sensor which is configurable it would be better.

Yes, that is one option, and I agree that it would be simpler. However, this is my first hacking attempt with homeassistant and I have a plan for some other things  I want to implement. So even if it would be a steep learning curve I am willing to take a shot at the more advanced approach.

epenet commented:
We have a bit of time before 2022.02 cut-off, so that would be great - the same will then be applicable for humidity sensors in `DS2438`, or temperature precision in `DS18B20`.

garbled1 commented:
If we add all 8, the user will have to go into HA and disable 7 of them, because they will display nonsense.  It seems like what you want is some kind of config type entity that defaults to K but lets you select from the set of letters, but do we even have an entity like that in HA? A multi-position switch?

epenet commented:
There is the `select` entity for this. But I'm not sure how to store and re-use that value.

droberg commented:
@garbled1 , there is an option in the setting that can default to disabled. This is used for example for the "volt", "typeK" and other entities. So instead of manually disabling we are currently defaulting to disabled and have to manually enable these entities.

garbled1 commented:
You mean in the setting on the device inside owserver?  I'm just trying to avoid adding 16 entities for every thermocouple...

droberg commented:
I'm currently digging around in some examples and trying to figure this out. I find it hard to believe that there would be no way of setting a configuration option on a device from a drop-down list of available alternatives. Currently the idea is to change it to a common thermocouple type instead of "typeK" in the entity-id. The drawback of this approach is that if someone changes the physical thermocouple type and the setting the id would be the same. But on the other hand, this is no different than if someone changes the usage of any other sensor by moving them or so...

droberg commented:
> You mean in the setting on the device inside owserver? I'm just trying to avoid adding 16 entities for every thermocouple...

No, not in owserver. So this alternative would result in a lot of unusable entities. They would be disabled by default, but still visible and clutter the view. I agree that this is not particularly nice.

epenet commented:
There might be a way to implement a configuration "select" entity which stores its value within home assistant, and then the sensors would read the configuration value from within home assistant.
All the examples I have found seem to store the configuration value outside HA though.

droberg commented:
How about the "Icon" and "Name" and "Entity ID" that is available as configurations as default for the entities. Are these special you think? The icon has a drop-down so i am looking around in the base classes to see how that is implemented.

garbled1 commented:
They aren't "special", but we really shouldn't attempt to override thier use with something else, that's way beyond unintuitive.  If nothing else works, I'd be ok with having all 8 types, and set them up as disabled by default, and note in the docs that you need to go enable the correct one, but only if thats the only way.

droberg commented:
There is also the option of using some of the memory on the chip to store the type of thermocouple. There are 2 pages with 16 bytes of user stored memory on these devices. This memory can be used for applications, such as a battery monitor, or similar to  store permanent values on config but we could use it for this setting. 

Because these devices can be used for multiple purposes, such as battery monitoring, thermocouples, and possibly other voltage sensing applications we cannot be sure that there is a connected thermocouple and that no one is using the memory. We could possibly have a global option enabling the use of the device memory for all such devices, thus assuming all connected chips are thermocouple converters. On the other hand, owfs seems to treat the device as a thermocouple converter with some added voltage and currnent sensing. 

While I can't see a use where someone would connect their devices to their own 1-wire bus if it's part of some complicated application I feel a little uncomfortable using this memory for all connected chips without asking. A global setting feels slightly better but not perfect. A device/entity setting would take us back to the original problem.

droberg commented:
> 

Ah, this is not what I meant. I was simply thinking that I could reverse-engineer their implementation

droberg commented:
Never mind, I just read that this is also used in the WS603 (also OWFS supported) weather station. This weather station seems to use the memory for calibration if I read the documentation correct.

droberg commented:
I had a look at the Unifi integration which looks promising. There is a config option in the integration config that depends on the attached network devices. 
![image](https://user-images.githubusercontent.com/62830707/148084312-bc18c3e1-d475-4a8f-a842-d1c3b8d2a861.png)

epenet commented:
May I suggest that you create a separate PR for the config-flow to ease the review?
Maybe a simple flow to choose between  `temperature` (=default) | `temperature9` | `temperature10` | `temperature11` for DS18B20 (device family 28)
Or a flow to choose between `humidity` (=default) | `HIH3600/humidity` | `HIH4000/humidity` | `HIH5030/humidity` | `HTM1735/humidity` for DS2438 (device family 26)

After that one is approved, then we can look back at this one with the selector for `B | E | J | K | N | R | S | T`

droberg commented:
Yes, the DS18B20 seems like a good starting point.

garbled1 commented:
won't that change it for all devices of that type?  While that might be ok for the temperature, because thats a precision, for humidity, that's problematic if you have 3 of different chip types, and for these thermocouples, it's assuming "all my thermocouples are identical"

epenet commented:
I think if possible it should be a per-device setting somehow.

garbled1 commented:
Right, I'm saying you can't do that in config flow configuration, afaik.

droberg commented:
It does look possible... The end result could end up being some kind  of hack, but It looks doable. 

The config flow for the Unifi integration implements a device-scan which adds devices to a  multi-select. So the options flow can be made dependent on the found devices as it seems. The actual configuration result is stored in a string listing the selected devices. I would say that it should be possible to something similar, i.e. create an options flow on-the fly and store the result in a string. Perhaps this string could be a dictionary or some someting json-encoded. 

It may be that this violates some rule I don't know of. It may also be that this does not work out in the end due some missing functionality with enums in the ui. The Unifi case shows that everything that is needed for a boolean selector is there. I have no idea how large a config string may be, but again, the Unifi case shows that it's not unfeasible. 

Worst case would be that the user has to provide the config-string manually, possibly by a little help from the ui listing the found devices. I am confident enough to give it a try, and even if it fails I have at least learnt something.

epenet commented:
Great I'll happily review. Don't hesitate to share it early.

epenet commented:
I am marking this one as draft until the config-flow amends are available.

droberg commented:
Fine by me!

I have been away for the weekend but I am still working on the config-flow. I was struggling a bit to figure out the inner workings of the translation stuff. I seems like there may be a problem adding text that are not pre-defined in the json file. My dynamically generated check boxes that I use for testing shows up, but I cannot give them a text that is not pre-defined as it seems. 

Possibly there will be some code to look at this week.

epenet commented:
You need to add the string to `strings.json` (compulsory) and to `translations\en.json` (for testing purposes).
Only the english language file is allowed in a PR - all other languages are handled automatically by the translator.
Finally, you might need to clear the browser cache as you are making the amends.

droberg commented:
Yes, I figured that out. However, my attempt on making a config-flow with dynamically generated stuff breaks that setup because I need to have a name for the sensor i the config-flow. That name cannot be in the json-file because it does not exist yet.

epenet commented:
Maybe what you need is description_placeholders: see https://github.com/home-assistant/core/blob/18706e6408b945c95d8c01b845a52a9da78f058a/homeassistant/components/renault/config_flow.py#L134-L140

jesserockz commented:
https://github.com/home-assistant/architecture/blob/master/adr/0019-GPIO.md

epenet commented:
> https://github.com/home-assistant/architecture/blob/master/adr/0019-GPIO.md

This doesn't apply to Owserver implementation (only Sysbus)

droberg commented:
There is now a draft in #64315 

Note that the draft only takes care of the actual config, not the actual usage of the config-parameter.

epenet commented:
Note: I have opened an arch discussion to see what the best way to solve this is: https://github.com/home-assistant/architecture/discussions/713

epenet commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -185,6 +185,36 @@ class OneWireSensorEntityDescription(OneWireEntityDescription, SensorEntityDescr
     "28": (SIMPLE_TEMPERATURE_SENSOR_DESCRIPTION,),
+    "30": (
+        SIMPLE_TEMPERATURE_SENSOR_DESCRIPTION,
+        OneWireSensorEntityDescription(
+            key="typeK/temperature",
+            device_class=SensorDeviceClass.TEMPERATURE,
+            entity_registry_enabled_default=False,
+            name="Temperature TypeK",
+            native_unit_of_measurement=TEMP_CELSIUS,
+            read_mode=READ_MODE_FLOAT,
+            state_class=SensorStateClass.MEASUREMENT,
+        ),
```
```suggestion
        OneWireSensorEntityDescription(
            key="type/temperature",
            device_class=SensorDeviceClass.TEMPERATURE,
            entity_registry_enabled_default=False,
            name="Temperature Type",
            native_unit_of_measurement=TEMP_CELSIUS,
            override_key="typeK/temperature",
            read_mode=READ_MODE_FLOAT,
            state_class=SensorStateClass.MEASUREMENT,
        ),
```

epenet replied:
I probably like `key="typeX/temperature"` best, because it matches the OWFS documentation.
Name could be adjusted also, to `TC temperature` or `Thermocouple temperature`, or...

epenet reviewed (commented):
I suggest that you create a new optional property in `OneWireSensorEntityDescription`, for example `overrride_key: str | None = None`.

Then when you create the device_file variable to pass it to the constructor, you use the overrride_key if it is available:
`os.path.join(...,description.override_key or description.key)`

If we do this, then we shouldn't need to wait for the config_flow PR to be merged.
We can then adjust the override_key property after the config_flow PR is merged, and it will not cause a breaking change.

droberg commented:
> I suggest that you create a new optional property in `OneWireSensorEntityDescription`, for example `overrride_key: str | None = None`.
> 

In the config flow for DS18B20 that I am working on this is almost exactly the solution I am going for. It's almost done but I didn't have time to push it yet. However, instead of 
`description.override_key or description.key` I have been using `description.key + description.extra_key` during the test. Yours is better though.

droberg commented on tests/components/onewire/const.py:
```diff
@@ -618,6 +618,58 @@
+                ATTR_UNIT_OF_MEASUREMENT: TEMP_CELSIUS,
+            },
+            {
+                ATTR_DEFAULT_DISABLED: True,
+                ATTR_DEVICE_CLASS: SensorDeviceClass.TEMPERATURE,
+                ATTR_ENTITY_ID: "sensor.30_111111111111_temperature_typek",
+                ATTR_INJECT_READS: b"    173.7563",
+                ATTR_STATE: "173.8",
+                ATTR_STATE_CLASS: SensorStateClass.MEASUREMENT,
+                ATTR_UNIQUE_ID: "/30.111111111111/typeK/temperature",
+                ATTR_UNIT_OF_MEASUREMENT: TEMP_CELSIUS,
```
Agreed

epenet commented:
Last day before cutoff for 2022.2 - `config_flow` won't be ready but maybe this one will?

droberg commented:
I will give it a shot

homeassistant commented:
Hello @droberg,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`a6f2bd60b106436108a22463e686db15fdb9d4f5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;63222&#x2F;commits&#x2F;a6f2bd60b106436108a22463e686db15fdb9d4f5) - This commit has something that looks like an email address (pero@wolfram.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

droberg committed:
Add support for DS2760 1-Wire TypeK Thermocouple

droberg commented:
Just a rebase to current devel

droberg committed:
Add support for changing of thermocouple type through override_key

droberg commented:
The test does not pass... I cant figure out why right now

epenet commented on tests/components/onewire/const.py:
```diff
@@ -618,6 +619,59 @@
+                ATTR_DEFAULT_DISABLED: True,
+                ATTR_DEVICE_CLASS: SensorDeviceClass.TEMPERATURE,
+                ATTR_DEVICE_FILE: "/30.111111111111/typeK/temperature",
+                ATTR_ENTITY_ID: "sensor.30_111111111111_temperature_typeX",
```
```suggestion
                ATTR_ENTITY_ID: "sensor.30_111111111111_thermocouple_temperature",
```

droberg replied:
Oh, yes. Checking again now

droberg committed:
Fix broken test for DS2760 Thermocouple

epenet commented on tests/components/onewire/const.py:
```diff
@@ -44,6 +44,7 @@
 ATTR_INJECT_READS = "inject_reads"
 ATTR_UNIQUE_ID = "unique_id"
 ATTR_UNKNOWN_DEVICE = "unknown_device"
+ATTR_OVERRIDE_KEY = "override_key"
```
This is not needed:
```suggestion
```

epenet reviewed (commented):
One last tweak, and I think this will be ready for review

droberg committed:
Remove unnecessary test-attribute for DS2760 Thermocouple

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

epenet reviewed (approved):
LGTM

epenet commented:
Thanks @droberg 👍

droberg commented:
Thank you @epenet !
